### PR TITLE
[SPARK-17654] [SQL] Enable populating hive bucketed tables

### PR DESCRIPTION
**Pull request by tejasapatil**
### Semantics:
- If the Hive table is bucketed, then INSERT node expect the child distribution to be based on the hash of the bucket columns. Else it would be empty. (Just to compare with Spark native bucketing : the required distribution is not enforced even if the table is bucketed or not... this saves the shuffle in comparison with hive).
- Sort ordering for INSERT node over Hive bucketed table is determined as follows:

| Table type   | Normal table | Bucketed table |
| ------------- | ------------- | ------------- |
| non-partitioned insert  | Nil | sort columns |
| static partition   | Nil | sort columns |
| dynamic partitions   | partition columns | (partition columns + bucketId + sort columns) |

Just to compare how sort ordering is expressed for Spark native bucketing:

| Table type   | Normal table | Bucketed table |
| ------------- | ------------- | ------------- |
|  sort ordering | partition columns | (partition columns + bucketId + sort columns) |

Why is there a difference ? With hive, since there bucketed insertions would need a shuffle, sort ordering can be relaxed for both non-partitioned and static partition cases. Every RDD partition would get rows corresponding to a single bucket so those can be written to corresponding output file after sort. In case of dynamic partitions, the rows need to be routed to appropriate partition which makes it similar to Spark's constraints.

- Only `Overwrite` mode is allowed for hive bucketed tables as any other mode will break the bucketing guarantees of the table. This is a difference wrt how Spark bucketing works.
- With the PR, if there are no files created for empty buckets, the query will fail. Will support creation of empty files in coming iteration. This is a difference wrt how Spark bucketing works as it does NOT need files for empty buckets.

### Summary of changes done:
- `ClusteredDistribution` and `HashPartitioning` are modified to store the hashing function used.
- `RunnableCommand`'s' can now express the required distribution and ordering. This is used by `ExecutedCommandExec` which run these commands
  - The good thing about this is that I could remove the logic for enforcing sort ordering inside `FileFormatWriter` which felt out of place. Ideally, this kinda adding of physical nodes should be done within the planner which is what happens with this PR.
- `InsertIntoHiveTable` enforces both distribution and sort ordering
- `InsertIntoHadoopFsRelationCommand` enforces sort ordering ONLY (and not the distribution)
- Fixed a bug due to which any alter commands to bucketed table (eg. updating stats) would wipe out the bucketing spec from metastore. This made insertions to bucketed table non-idempotent operation.

## How was this patch tested?

- Added new unit tests

### CODE REVIEW

**Comment by tejasapatil** [comment]
Jenkins test this please

**Comment by tejasapatil** [comment]
Jenkins retest this please

**Comment by tejasapatil** [comment]
cc @cloud-fan @gatorsmile @sameeragarwal @rxin

**Comment by tejasapatil** [comment]
Jenkins retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 1]
```diff
@@ -50,7 +50,9 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
       numPartitions: Int): Partitioning = {
     requiredDistribution match {
       case AllTuples => SinglePartition
-      case ClusteredDistribution(clustering) => HashPartitioning(clustering, numPartitions)
+      case ClusteredDistribution(clustering, numClusters, hashingFunctionClass) =>
+        HashPartitioning(clustering, numClusters.getOrElse(numPartitions),
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: This is going to create a partitioning that satisfies that distribution. According to modified  `HashPartitioning`, if `numPartitions` isn't equal to `numClusters`, `satisfies` returns `false`. It seems a conflict if we ask to create a partitioning of `numPartitions` with a `ClusteredDistribution` of `numClusters`?

Comment (tejasapatil) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: Good point. I gave this more thought and have made changes

**Comment by tejasapatil** [comment]
I have a new PR (https://github.com/apache/spark/pull/19001) which supersedes this one. It has everything this PR does (ie. writer side changes) plus reader side changes.

**Comment by maropu** [comment]
@tejasapatil Can you close this for now because it's not active for a long time.

**Pull request closed without merging**